### PR TITLE
[JAX][DeepSeek/MoE] Create `UnquantizedFusedMoEBackend` (sans weight loading)

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

In this PR, I implement the core logic for the `UnquantizedMoEBackend` while keeping reverse compatibility (for now) for the Megablocks + Qwix quantized path.   Specifically, I do the following:

* Move all of the MoE forward pass logic to be contained in `UnquantizedFusedMoEBackend.`. The `FP8UnquantizedFusedMoEBackend` should be landing soon to provide full quantization support.
* Wire up the `DENSE_MAT` and `MEGABLOCKS_GMM` paths in `moe_apply`
* Refactor sparse/dense MoE to be self-contained
* Move backend-specific MoE weight logic to `UnquantizedFusedMoEMethod` from the `JaxMoE` class
* Add `process_weights_after_loading` as a (potentially inheritable) method on `QuantizeMethodBase`
* Implement `apply_jax`, `create_jax_weights`, and `process_weights_after_loading` on `UnquantizedFusedMoEBackend`
* Add a `_process_weights_after_loading` function to `model_loader.py` to call `_process_weights_after_loading`

# Tests

Ensured offline inference outputs matched between `main` for all combinations of the following:
* MHA/MLA
* GMM_TP and MEGABLOCKS_GMM
* Quantization (only for MEGABLOCKS_GMM)

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/jax/moe/moe.py [hunk 1]
```diff
@@ -321,96 +191,40 @@ def __post_init__(self, rngs: nnx.Rngs):
...
-                dtype=self.dtype,
-                sharding=self.efd_sharding,
-                random_init=self.random_init)
-
-        # Default MoE has no bias vectors
-        self.w1_bias, self.w2_bias = (None, None)
-
-        # TODO: Add quantization scale params for VLLM MoE kernel
-        self.w1_scale, self.w2_scale = (None, None)
+        # TODO (jacobplatin): we need to clean up the shardings, but the issue is that
```
Comment (jrplatin) [inline tpu_inference/layers/jax/moe/moe.py]: Want to address this at some point soon

File: tpu_inference/layers/jax/quantization/__init__.py [hunk 2]
```diff
@@ -35,6 +36,10 @@ def get_tpu_quantization_config(vllm_config: VllmConfig):
         # TODO(#1623): replace with actual FP8 config when delegation is ready.
         FP8: lambda: None,
     }
+    # TODO (jacobplatin): temporary workaround for now before FP8 is fully ready for DeepSeek
```
Comment (jrplatin) [inline tpu_inference/layers/jax/quantization/__init__.py]: Very temporary hack

File: tpu_inference/layers/jax/quantization/unquantized.py [hunk 3]
```diff
@@ -49,18 +54,120 @@ def apply_jax(self, layer: JaxModule, x: jax.Array) -> jax.Array:
...
+                "bf": 512,
+                "bd1": 512,
+                "bd2": 512,
+                "btc": 64,
+                "bfc": 256,
+                "bd1c": 256,
+                "bd2c": 256,
+            }
+
+            # TODO (jacobplatin): why is this sharding flipped?
```
Comment (jrplatin) [inline tpu_inference/layers/jax/quantization/unquantized.py]: @bzgoogle  I'd like to understand why this is the case

File: tpu_inference/models/jax/deepseek_v3.py [hunk 4]
```diff
@@ -1533,6 +1471,13 @@ def __init__(self,
         self.vllm_config = vllm_config
         self.rng = nnx.Rngs(rng)
 
+        # TODO (jacobplatin): we will resolve this issue in a forthcoming PR that will refactor weight loading
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: Note

File: tpu_inference/layers/jax/quantization/unquantized.py [hunk 5]
```diff
@@ -49,18 +54,119 @@ def apply_jax(self, layer: JaxModule, x: jax.Array) -> jax.Array:
...
+
+        router_logits = None
+        # Fused weight backends
+        if layer.moe_backend in MoEBackend.fused_moe_backends():
+            # of shape TE, only 1D in this case
+            router_logits = layer.router(x_TD)
+
+            w13_weight = layer.kernel_gating_upproj_E2DF.value if layer.moe_backend == MoEBackend.FUSED_MOE else layer.kernel_gating_upproj_EDF.value
+            w2_weight = layer.kernel_down_proj_EFD.value
+            # TODO (jacobplatin/bzgoogle): we should support bias
```
Comment (jrplatin) [inline tpu_inference/layers/jax/quantization/unquantized.py:133]: Note

File: tpu_inference/envs.py [hunk 6]
```diff
@@ -152,6 +153,8 @@ def _get_bool_env() -> bool:
     "USE_MEGABLOCKS":
     env_bool("USE_MEGABLOCKS", default=False),
     # Number of TPU slices for multi-slice mesh
+    "USE_DENSE_MOE":
+    env_bool("USE_DENSE_MOE", default=False),
```
Comment (lk-chen) [inline tpu_inference/envs.py:159]: Could you add comment like other envs did?

File: tpu_inference/models/jax/deepseek_v3.py (hunk 4 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: you can use #1623

File: tpu_inference/layers/jax/quantization/unquantized.py [hunk 7]
```diff
@@ -49,18 +54,119 @@ def apply_jax(self, layer: JaxModule, x: jax.Array) -> jax.Array:
 
 class UnquantizedFusedMoEMethod(QuantizeMethodBase):
     """
-    TODO (jacobplatin): implement in forthcoming PR.
+    Unquantized method for JAXMoE layer.
+
+    TODO (jacobplatin): support weight loading -- currently, model-dependent.
```
Comment (lk-chen) [inline tpu_inference/layers/jax/quantization/unquantized.py:59]: same here, you can use #1623

File: tpu_inference/models/common/model_loader.py [hunk 8]
```diff
@@ -96,6 +96,21 @@ def _get_nnx_model(
     rng: jax.Array,
     mesh: Mesh,
 ) -> nnx.Module:
+    vllm_config.quant_config = get_tpu_quantization_config(vllm_config)
+
+    def _process_weights_after_loading(model: nnx.Module) -> None:
```
Comment (jrplatin) [inline tpu_inference/models/common/model_loader.py]: Looks like https://github.com/vllm-project/tpu-inference/pull/1611/changes#diff-68dc98fbd69cf60fc9de428c6fb117a12bfd17c70bc13985885f7e2bdec0ee36R573

Is also implementing this, so will need to coordinate with @patemotter

File: tpu_inference/layers/jax/quantization/unquantized.py [hunk 9]
```diff
@@ -49,18 +54,119 @@ def apply_jax(self, layer: JaxModule, x: jax.Array) -> jax.Array:
 
 class UnquantizedFusedMoEMethod(QuantizeMethodBase):
```
Comment (lk-chen) [inline tpu_inference/layers/jax/quantization/unquantized.py:55]: Comparing 2 pattern:
1. There's one UnquantizedFusedMoEMethod, switching to different branches according to `moe_backend`, call `moe_apply` which is still branching according to `moe_backend`
2. Implement `UnquantizedGmmMoEMethod` `UnquantizedMegabloxGmmMoEMethod` separately and call `gmm_moe_apply` `megablox_gmm_moe_apply` each

Is there a reason behind why you choose 1?

File: tpu_inference/layers/jax/quantization/__init__.py (hunk 2 above)
Comment (lk-chen) [reply tpu_inference/layers/jax/quantization/__init__.py]: is it intended to check in?

File: tpu_inference/layers/common/moe.py [hunk 10]
```diff
@@ -124,11 +124,32 @@ def moe_apply(
...
                 )
             case MoEBackend.DENSE_MAT:
-                # TODO(jacobplatin): will implement in forthcoming PR
-                raise NotImplementedError
+                # NOTE: circular import avoidance
+                from tpu_inference.layers.jax.moe.dense_moe import \
+                    dense_moe_func
+                assert len(
+                    gating_output
+                ) == 2, "Expected the gating output to be have 2 entries: weights and indices"
```
Comment (patemotter) [inline tpu_inference/layers/common/moe.py:135]: Based on the type hints this can be a jax.Array as well as a tuple. Will this work for both cases? I believe len(jax.Array) will return the size of the first dim.

File: tpu_inference/layers/jax/quantization/unquantized.py (hunk 9 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/quantization/unquantized.py:55]: 1 is more in-line with what [TorchAX](https://github.com/vllm-project/tpu-inference/blob/main/tpu_inference/layers/common/moe.py#L59) is currently doing (one Method with the multiple backends called via `moe_apply`) so figured I'd keep things consistent

File: tpu_inference/layers/jax/quantization/__init__.py (hunk 2 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/quantization/__init__.py]: Yes, I think so -- otherwise, DeepSeek can't be loaded with the refactored code, but as I noted, this is very temporary and should be removed as soon as FP8 support is done (hopefully by EOW)

Comment (kyuyeunk) [reply tpu_inference/layers/jax/quantization/__init__.py]: can you share why deepseek can't be loaded with the refactored code?

File: tpu_inference/layers/jax/moe/dense_moe.py [hunk 11]
```diff
@@ -11,54 +11,139 @@
...
+
+
+def dense_moe_func(weights: UnfusedMoEWeights, x_TD: jax.Array,
+                   gating_output: Tuple[jax.Array, jax.Array],
+                   cast_dtype: jnp.dtype, num_local_experts: int,
+                   apply_expert_weight_before_computation: bool,
+                   activation_ffw_td, hidden_act) -> jax.Array:
+    """
+    Forward pass of the dense MoE layer.  This is a naive implementation
+    and thus should not be used in production.
```
Comment (patemotter) [inline tpu_inference/layers/jax/moe/dense_moe.py:105]: Nit: do we print an explicit logging warning about this anywhere?

File: tpu_inference/layers/jax/quantization/unquantized.py [hunk 12]
```diff
@@ -49,18 +54,119 @@ def apply_jax(self, layer: JaxModule, x: jax.Array) -> jax.Array:
...
+        """
+        if layer.moe_backend == MoEBackend.FUSED_MOE:
+            if layer.edf_sharding:
+                e2df_sharding = (layer.edf_sharding[0], None,
+                                 layer.edf_sharding[1], layer.edf_sharding[2])
+            # fuse the weights into w13: [Gate, Up]
+            w_gate = layer.kernel_gating_EDF.value
+            w_up = layer.kernel_up_proj_EDF.value
+
+            # stack to create a 4d matrix
```
Comment (patemotter) [inline tpu_inference/layers/jax/quantization/unquantized.py]: The most pedantic nit: 4d tensor or 4d array would be more correct.

File: tpu_inference/layers/jax/quantization/__init__.py (hunk 2 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/quantization/__init__.py]: Sorry, what I said above wasn't super clear -- when I say "refactored code" what I mean is the "refactored code in its current state".  This is a very temporary hack that we'll remove once we actually have the FP8 support, but it just enables us to run the quantized model between now and then (when the refactor is actually done)

File: tpu_inference/layers/jax/moe/dense_moe.py (hunk 11 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/moe/dense_moe.py:105]: I can add a `warning_once`!

File: tpu_inference/envs.py (hunk 6 above)
Comment (jrplatin) [reply tpu_inference/envs.py:159]: Added!

File: tpu_inference/layers/common/moe.py (hunk 10 above)
Comment (jrplatin) [reply tpu_inference/layers/common/moe.py:135]: Ah what a catch, thanks Pate -- updated!

File: tpu_inference/envs.py [hunk 13]
```diff
@@ -21,7 +21,8 @@
     PHASED_PROFILING_DIR: str = ""
     PYTHON_TRACER_LEVEL: int = 1
     USE_MOE_EP_KERNEL: bool = False
-    USE_MEGABLOCKS: bool = False
+    USE_UNFUSED_MEGABLOCKS: bool = False
```
Comment (jrplatin) [inline tpu_inference/envs.py:24]: I renamed this BTW because I realized `USE_MEGABLOCKS` is too vague -- I _think_ this is a clear name but open to feedback obviously!

File: tpu_inference/layers/jax/quantization/unquantized.py (hunk 9 above)
Comment (lk-chen) [reply tpu_inference/layers/jax/quantization/unquantized.py:55]: Got it. For future reference vLLM use if-branch's to create weight [here](https://github.com/vllm-project/vllm/blob/92e7562a994038c904fea859d90462c7e84a3246/vllm/model_executor/layers/quantization/mxfp4.py#L250-L274)

File: tpu_inference/layers/jax/quantization/__init__.py (hunk 2 above)
Comment (lk-chen) [reply tpu_inference/layers/jax/quantization/__init__.py]: IIUC, DS-R1 needs FP8 einsum to fully work, which is missing right now.

Image

Could you explain (maybe in the TODO comment) if we load DS-R1-FP8 checkpoint, how does it not hit exception at einsum? If there's exception, we should remove this hack (you can have local patch for dev.)

Comment (jrplatin) [reply tpu_inference/layers/jax/quantization/__init__.py]: Ah so this branch was based on the current `main` yesterday where the `UnquantizedLinearMethod` wasn't hooked up yet -- if that has been hooked up, then I think you're right in that we'll get issues, but let me double check.  We might be ok because DeepSeek isn't using `JaxLinear` at all anyways

Comment (jrplatin) [reply tpu_inference/layers/jax/quantization/__init__.py]: I moved this logic to the DeepSeek modeling file for now, thanks @kyuyeunk !

File: tpu_inference/models/common/model_loader.py (hunk 8 above)
Comment (jrplatin) [reply tpu_inference/models/common/model_loader.py]: I took over `_process_modules_after_loading` so it should be easy to merge, regardless of who goes first

File: tpu_inference/models/jax/utils/weight_utils.py [hunk 14]
```diff
@@ -568,6 +568,22 @@ def load_hf_weights(
     nnx.update(model, params)
 
 
+def process_modules_after_loading(module, mesh):
```
Comment (lk-chen) [inline tpu_inference/models/jax/utils/weight_utils.py]: nit: similar to out discussion regarding the config, if this is temporary for ds-v3, you may want to move it to ds_v3.py to control the exclosure of this function

Comment (lk-chen) [inline tpu_inference/models/jax/utils/weight_utils.py]: nit: similar to out discussion regarding the config, if this is temporary for ds-v3, you may want to move it to ds_v3.py to control the exclosure of this function

File: tpu_inference/models/common/model_loader.py [hunk 15]
```diff
@@ -96,6 +96,7 @@ def _get_nnx_model(
     rng: jax.Array,
     mesh: Mesh,
 ) -> nnx.Module:
+    vllm_config.quant_config = get_tpu_quantization_config(vllm_config)
```
Comment (lk-chen) [inline tpu_inference/models/common/model_loader.py]: Line 245 already has this

File: tpu_inference/layers/jax/quantization/__init__.py [hunk 16]
```diff
@@ -27,6 +27,7 @@ def get_tpu_quantization_config(vllm_config: VllmConfig):
         UnquantizedConfig
 
     model_config = copy.deepcopy(vllm_config.model_config)
+    # TODO: add fp8
```
Comment (lk-chen) [inline tpu_inference/layers/jax/quantization/__init__.py]: Line 36 should be sufficient

File: tpu_inference/models/jax/utils/weight_utils.py (hunk 14 above)
Comment (jrplatin) [reply tpu_inference/models/jax/utils/weight_utils.py]: I think Pate has this logic too no? So if it's needed for both of us, I figured just add it in the shared weight utils

Comment (jrplatin) [reply tpu_inference/models/jax/utils/weight_utils.py]: Ah didn't see [this](https://github.com/vllm-project/tpu-inference/pull/1611/changes#r2766542566) -- so this will be removed?

File: tpu_inference/models/common/model_loader.py (hunk 15 above)
Comment (jrplatin) [reply tpu_inference/models/common/model_loader.py]: Ah good catch -- rebased and missed thanks!

File: tpu_inference/models/jax/utils/weight_utils.py (hunk 14 above)
Comment (jrplatin) [reply tpu_inference/models/jax/utils/weight_utils.py]: Moved!

File: tpu_inference/layers/jax/moe/dense_moe.py [hunk 17]
```diff
@@ -11,54 +11,139 @@
...
+                              axis=1)
+    # Some models use the routing scores to weight the data instead of
+    # weighting the expert outputs.
+    if apply_expert_weight_before_computation:
+        with jax.named_scope("pre_computing_weight"):
+            return dense_moe_fwd_preapply_router_weights(
+                weights=weights,
+                x_TD=x_TD,
+                cast_dtype=cast_dtype,
+                activation_ffw_td=activation_ffw_td,
```
Comment (patrickji2014) [inline tpu_inference/layers/jax/moe/dense_moe.py:140]: there is inconsistency in calling dense_moe_fwd_preapply_router_weights, activation_ffw_td is not an arg of the function, it should be activation_ffw_ted. either one need a fix

Comment (jrplatin) [reply tpu_inference/layers/jax/moe/dense_moe.py:140]: https://github.com/vllm-project/tpu-inference/pull/1685

Fix out here
